core/trino-main/src/main/java/io/trino/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -588,49 +621,300 @@ public Type getFinalType()
         }
 
         @Override
-        public Type getIntermediateType()
+        public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        {
+            groupCount = max(groupCount, groupIdsBlock.getGroupCount());
+            // Add group id block
+            pagesIndex.addPage(page.appendColumn(groupIdsBlock));
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            accumulator.evaluateFinal(groupId, output);
+        }
+
+        @Override
+        public void prepareFinal()
+        {
+            pagesIndex.sort(orderByChannels, orderings);
+            Iterator<Page> pagesIterator = pagesIndex.getSortedPages();
+            pagesIterator.forEachRemaining(page -> {
+                // The last channel of the page is the group id
+                GroupByIdBlock groupIds = new GroupByIdBlock(groupCount, page.getBlock(page.getChannelCount() - 1));
+                // We pass group id together with the other input channels to accumulator. Accumulator knows which input channels
+                // to use. Since we did not change the order of original input channels, passing the group id is safe.
+                accumulator.addInput(groupIds, page);
+            });
+        }
+
+        @Override
+        public final Type getIntermediateType()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public final void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public final void evaluateIntermediate(int groupId, BlockBuilder output)
         {
             throw new UnsupportedOperationException();
         }
+    }
+
+    /**
+     * {@link SpillableFinalOnlyGroupedAccumulator} enables spilling for an accumulator that does not support partial aggregation
+     */
+    private static class SpillableFinalOnlyGroupedAccumulator
```

Comment:
This doesn't seem like the right approach (wrapping the accumulators with this accumulator that knows how to produce and process intermediates). Instead, we should just update the DistinctAccumulator and OrderingAccumulator to be able to produce and take intermediates natively.

Reply:
How about the code duplication? It seems that the deleted abstract class "FinalOnlyGroupedAccumulator" needs to be introduced again in order to remove this wrapping.

Reply:
It wouldn't be duplicated code, since instead of synthesizing generic rows and pages, each accumulator would produce and process intermediates that make sense for the operation they perform. I.e., the intermediates would behave as if they were being used for a partial aggregation (whether we enable partials for distinct and order by is a separate discussion -- it may be worth doing, regardless)

Reply:
Done